### PR TITLE
104 ocr

### PR DESCRIPTION
**Pull request by sarge1989**
Changes
- Implemented OCR
- Significant changes to schema to accommodate image and text being able to cross-match
  - Importantly "text" and "caption" are now separate
  - We have a new "captionHash" to allow filtering of Typesense. A text message can match an image's OCR text only if captionHash is null
  - OCR text can match another image's OCR text only if the captionHash is the same
  - Removed "type" from message as a message can now contain either image or text
    - Decision to send image message or text to checkers is based on last message instead of message type
    - Basically to be in the same message, the "text" (includes extracted text) must be similar AND "caption" must also match
    - Fields "mimeType", "storageURL", "hash", "mediaId" all shifted to instances

### CODE REVIEW

File: functions/src/definitions/checkerHandlers.js [hunk 1]
```diff
@@ -195,8 +195,11 @@ async function onFactCheckerYes(voteRequestPath, from, platform = "whatsapp") {
   const messageRef = voteRequestRef.parent.parent
   const messageSnap = await messageRef.get()
   const message = messageSnap.data()
+  const latestInstanceRef = messageSnap.get("latestInstance")
+  const latestInstanceSnap = await latestInstanceRef.get()
+  const latestType = latestInstanceSnap.get("type") ?? "text"
   let res
-  switch (message.type) {
+  switch (latestType) {
```
Comment (stonefruit) [inline functions/src/definitions/checkerHandlers.js:202]: To prevent function from crashing, will be good to handle the `default` case. It looks like in our case, the logic for `text` can be used as default.

File: functions/src/definitions/common/machineLearningServer/operations.ts [hunk 2]
```diff
@@ -40,6 +54,14 @@ async function getL1Category(text: string): Promise<string> {
   return response.data.prediction
 }
 
+async function performOCR(url: string): Promise<OCRResponse> {
+  const data = {
+    url: url,
+  }
+  const response = await callAPI<OCRResponse>("ocr", data)
+  return response.data
```
Comment (stonefruit) [inline functions/src/definitions/common/machineLearningServer/operations.ts:62]: Will be good to map data to camelCase here since the returned value is using snake_case for consistency in other parts of the app

File: functions/src/definitions/common/mediaUtils.ts [hunk 3]
```diff
@@ -60,20 +63,24 @@ async function getHash(buffer: Buffer) {
...
+          action: "read",
+          expires: Date.now() + 60 * 60 * 1000,
+        })
+      return temporaryUrl
+    } catch (error) {
+      functions.logger.error(error)
+      // You can also log the error to an external monitoring service or send an alert to the developers
+      return null
+    }
+  } else {
```
Comment (stonefruit) [inline functions/src/definitions/common/mediaUtils.ts:82]: To reduce the depth of the prod code, moving this up to early return can be used `if (runtimeEnvironment.value() !== "PROD") {` can reduce the use one 1 set of {}

File: functions/src/definitions/onInstanceCreate.js [hunk 4]
```diff
@@ -68,6 +69,28 @@ exports.onInstanceCreate = functions
...
+      })
+      if (data?.embedding && data?.text) {
+        const updateObj = {
+          id: snap.ref.path.replace(/\//g, "_"), //typesense id can't seem to take /
+          message: data.text,
+          captionHash: data.captionHash ? data.captionHash : "__NULL__",
+          embedding: data.embedding,
+        }
+        try {
+          await insertOne(updateObj, CollectionTypes.Instances)
```
Comment (stonefruit) [inline functions/src/definitions/onInstanceCreate.js:86]: repeated logic, whole block could be inside 1 try/catch actually

File: functions/src/definitions/userHandlers.js [hunk 5]
```diff
@@ -275,96 +287,194 @@ async function newImageInstanceHandler({
...
+    stream.on("finish", resolve)
+    stream.end(buffer)
+  })
+  let matchType = "none" // will be set to either "similarity" or "image" or "none"
+  //check if identical image exists among instances
   let imageMatchSnapshot = await db
-    .collection("messages")
-    .where("type", "==", "image")
+    .collectionGroup("instances")
     .where("hash", "==", hash)
```
Comment (stonefruit) [inline functions/src/definitions/userHandlers.js:312]: This seems to be using AND logic. I would assume that the image hash alone matching would be enough to consider it a match? or is it intended that image + caption pairing is always required?

File: functions/src/definitions/userHandlers.js [hunk 6]
```diff
@@ -275,96 +287,194 @@ async function newImageInstanceHandler({
...
   //get response buffer
   let buffer = await downloadWhatsappMedia(mediaId, mimeType)
   const hash = await getHash(buffer)
+
+  //upload to storage
+  const storageBucket = admin.storage().bucket()
+  filename = `images/${mediaId}.${mimeType.split("/")[1]}`
+  const file = storageBucket.file(filename)
+  const stream = file.createWriteStream()
+  await new Promise((resolve, reject) => {
```
Comment (stonefruit) [inline functions/src/definitions/userHandlers.js:303]: Hash can be checked before uploading to avoid having the storage grow

**Review by stonefruit — CHANGES_REQUESTED** [review]

File: functions/src/definitions/userHandlers.js (hunk 5 above)
Comment (sarge1989) [reply functions/src/definitions/userHandlers.js:312]: Intended, cos the caption may change the meaning of the message.

File: functions/src/definitions/checkerHandlers.js (hunk 1 above)
Comment (sarge1989) [reply functions/src/definitions/checkerHandlers.js:206]: Done

File: functions/src/definitions/common/machineLearningServer/operations.ts (hunk 2 above)
Comment (sarge1989) [reply functions/src/definitions/common/machineLearningServer/operations.ts]: Done

File: functions/src/definitions/common/mediaUtils.ts (hunk 3 above)
Comment (sarge1989) [reply functions/src/definitions/common/mediaUtils.ts]: Done

File: functions/src/definitions/onInstanceCreate.js (hunk 4 above)
Comment (sarge1989) [reply functions/src/definitions/onInstanceCreate.js]: Done

File: functions/src/definitions/userHandlers.js (hunk 6 above)
Comment (sarge1989) [reply functions/src/definitions/userHandlers.js]: Done

**Review by stonefruit — APPROVED** [review]
Nice. Thanks for your updates.
